pmlc/target/x86/pipeline.cc

```diff
@@ -104,25 +84,124 @@ struct ConvertStandardToLLVMPass
     conversion::stdx_to_llvm::populateStdXToLLVMConversionPatterns(
         typeConverter, patterns);
     populateOpenMPToLLVMConversionPatterns(context, typeConverter, patterns);
-    patterns.insert<ExtractLoweringPattern>(context);
 
     LLVMConversionTarget target(*context);
     target.addDynamicallyLegalOp<omp::ParallelOp>([&](omp::ParallelOp op) {
       return typeConverter.isLegal(&op.getRegion());
     });
     target.addLegalOp<omp::TerminatorOp, omp::TaskyieldOp, omp::FlushOp,
                       omp::BarrierOp, omp::TaskwaitOp>();
-    target.addDynamicallyLegalOp<LLVM::ExtractValueOp>(
-        [](LLVM::ExtractValueOp op) {
-          return !mlir::dyn_cast_or_null<LLVM::InsertValueOp>(
-              op.container().getDefiningOp());
-        });
     if (failed(applyPartialConversion(module, target, patterns))) {
       signalPassFailure();
     }
   }
 };
 
+// OpenMP has issues passing values through to OpenMP blocks.  As a workaround,
+// we have a simple pass to smuggle values that cross the boundary via an
+// alloca()'d struct.
+struct OpenMPWorkaroundPass final
+    : public OpenMPWorkaroundBase<OpenMPWorkaroundPass> {
+  void runOnOperation() final {
+    LLVM::LLVMFuncOp funcOp = getOperation();
+    OpBuilder builder{&getContext()};
+    funcOp.walk([&](omp::ParallelOp parOp) {
+      // The values used to construct the smuggling struct.
+      SmallVector<LLVM::LLVMType, 8> fieldTypes;
+      SmallVector<Value, 8> fieldSrcs;
+
+      // A mapping from original values to the OpOperands within the
+      // parallel region that use those values, for value replacement.
+      std::unordered_map<void *, std::list<OpOperand *>> valueMap;
```

Comment:
We should never use `void*` for the key, what are we trying to do here?

Reply:
Duplicate detection.  Honestly, I just copied the pattern from MLIR's `BlockAndValueMapping`.